loadtest/control/simulcontroller/actions.go

```diff
@@ -1282,3 +1300,357 @@ func sendTypingEventIfEnabled(u user.User, channelId string) error {
+		})
+		if err != nil {
+			return control.UserActionResponse{Err: control.NewUserError(err)}
+		}
+		oldestUnreadThreadId = unreadThreads[len(unreadThreads)-1].PostId
+		// idle time between scrolls, between 1 and 10 seconds.
+		idleTime := time.Duration(1+rand.Intn(10)) * time.Second
+		select {
+		case <-c.stopChan:
+			return control.UserActionResponse{Info: "action canceled"}
+		case <-time.After(idleTime):
+		}
+	}
+
+	return control.UserActionResponse{Info: "Visited Global Threads Screen"}
+}
+
+func (c *SimulController) followThread(u user.User) control.UserActionResponse {
+	collapsedThreads, resp := control.CollapsedThreadsEnabled(u)
+	if resp.Err != nil || !collapsedThreads {
+		return resp
+	}
+	channel, err := u.Store().CurrentChannel()
+	if err != nil {
+		if errors.Is(err, memstore.ErrChannelNotFound) {
+			return control.UserActionResponse{Info: "followThread: current channel not set"}
+		}
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	post, err := u.Store().RandomPostForChannel(channel.Id)
+	if err != nil {
+		if errors.Is(err, memstore.ErrPostNotFound) {
+			return control.UserActionResponse{Info: "followThread: can't find post to follow in channel"}
+		}
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	try := 0
+	for try < 100 && post.RootId != "" {
+		post, err = u.Store().RandomPostForChannel(channel.Id)
+		if err != nil {
+			if errors.Is(err, memstore.ErrPostNotFound) {
+				return control.UserActionResponse{Info: "followThread: can't find post to follow in channel"}
+			}
+			return control.UserActionResponse{Err: control.NewUserError(err)}
+		}
+		try++
+	}
+	if post.RootId != "" {
+		return control.UserActionResponse{Info: "followThread: can't find post to follow in channel"}
+	}
+
+	err = u.UpdateThreadFollow(channel.TeamId, post.Id, true)
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	return control.UserActionResponse{Info: fmt.Sprintf("followed thread %s", post.Id)}
+}
+
+func (c *SimulController) unfollowThread(u user.User) control.UserActionResponse {
+	collapsedThreads, resp := control.CollapsedThreadsEnabled(u)
+	if resp.Err != nil || !collapsedThreads {
+		return resp
+	}
+	thread, err := u.Store().RandomThread()
+	if err != nil {
+		if errors.Is(err, memstore.ErrThreadNotFound) {
+			return control.UserActionResponse{Info: "unfollow thread: no thread to unfollow"}
+		}
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	channel, err := u.Store().Channel(thread.Post.ChannelId)
+	if err != nil || channel == nil {
+		err = u.GetChannel(thread.Post.ChannelId)
+		if err != nil {
+			return control.UserActionResponse{Err: control.NewUserError(err)}
+		}
+		channel, err := u.Store().Channel(thread.Post.ChannelId)
+		if err != nil || channel == nil {
+			return control.UserActionResponse{Err: control.NewUserError(errors.New("unfollow thread: can't get channel for thread"))}
+		}
```

Comment:
Let's separate these 2 cases. In the case of `err != nil`, we need to bubble up the error.